vllm/attention/backends/flash_attn.py

```diff
@@ -178,28 +178,49 @@ def decode_metadata(self) -> Optional["FlashAttentionMetadata"]:
         assert self.block_tables is not None
         assert self.seq_lens_tensor is not None
 
-        self._cached_decode_metadata = FlashAttentionMetadata(
-            num_prefills=0,
-            num_prefill_tokens=0,
-            num_decode_tokens=self.num_decode_tokens,
-            slot_mapping=self.slot_mapping[self.num_prefill_tokens:],
-            seq_lens=None,
-            seq_lens_tensor=self.seq_lens_tensor[self.num_prefills:],
-            max_query_len=None,
-            max_prefill_seq_len=0,
-            max_decode_seq_len=self.max_decode_seq_len,
-            query_start_loc=None,
-            seq_start_loc=None,
-            context_lens_tensor=None,
-            block_tables=self.block_tables[self.num_prefills:],
-            use_cuda_graph=self.use_cuda_graph,
-        )
+        if self.num_prefill_tokens == 0:
+            # This case avoids the tensor slicing overheads
+            self._cached_decode_metadata = FlashAttentionMetadata(
+                num_prefills=0,
+                num_prefill_tokens=0,
+                num_decode_tokens=self.num_decode_tokens,
+                slot_mapping=self.slot_mapping,
+                seq_lens=None,
+                seq_lens_tensor=self.seq_lens_tensor,
+                max_query_len=None,
+                max_prefill_seq_len=0,
+                max_decode_seq_len=self.max_decode_seq_len,
+                query_start_loc=None,
+                seq_start_loc=None,
+                context_lens_tensor=None,
+                block_tables=self.block_tables,
+                use_cuda_graph=self.use_cuda_graph,
+            )
+        else:
+            self._cached_decode_metadata = FlashAttentionMetadata(
+                num_prefills=0,
+                num_prefill_tokens=0,
+                num_decode_tokens=self.num_decode_tokens,
+                slot_mapping=self.slot_mapping[self.num_prefill_tokens:],
+                seq_lens=None,
+                seq_lens_tensor=self.seq_lens_tensor[self.num_prefills:],
+                max_query_len=None,
+                max_prefill_seq_len=0,
+                max_decode_seq_len=self.max_decode_seq_len,
+                query_start_loc=None,
+                seq_start_loc=None,
+                context_lens_tensor=None,
+                block_tables=self.block_tables[self.num_prefills:],
+                use_cuda_graph=self.use_cuda_graph,
+            )
```

Comment:
could de-dup a bit here by just setting those three values as vars in the if/else and then having a single call to the constructor?

Reply:
Thanks for the suggestion, applied this change in https://github.com/vllm-project/vllm/pull/7206